### PR TITLE
GH-116090: Fix test and clarify behavior for exception events when exhausting a generator.

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-116090

----
📚 Documentation preview 📚: https://cpython-previews--120697.org.readthedocs.build/

### CODE REVIEW

**Comment by markshannon** [comment]
Skipping news as nothing has changed here.

File: Doc/library/sys.monitoring.rst [hunk 1]
```diff
@@ -226,6 +226,10 @@ To allow tools to monitor for real exceptions without slowing down generators
 and coroutines, the :monitoring-event:`STOP_ITERATION` event is provided.
 :monitoring-event:`STOP_ITERATION` can be locally disabled, unlike :monitoring-event:`RAISE`.
 
+Note that the :monitoring-event:`STOP_ITERATION` event and the :monitoring-event:`RAISE`
+event for a :exc:`StopIteration` exception are equivalent, and are treated as interchangeable
+when generating events. Implementations will favor :monitoring-event:`STOP_ITERATION` for
+performance reasons, but may generate a RAISE event with a :exc:`StopIteration`.
```
Comment (gaogaotiantian) [inline Doc/library/sys.monitoring.rst]:
```suggestion
performance reasons, but may generate a :monitoring-event:`RAISE` event with a :exc:`StopIteration`.
```

File: Lib/test/test_monitoring.py [hunk 2]
```diff
@@ -832,20 +832,41 @@ def func1():
...
+            if iterator is None:
+                iterator = gen()
+            for _ in iterator:
                 pass
 
-        self.check_events(implicit_stop_iteration, [("raise", StopIteration)], recorders=(StopiterationRecorder,))
+        recorders=(ExceptionRecorder, StopiterationRecorder,)
+        expected = [("raise", StopIteration)]
+
+        # Make specialization fail, so that we can test the unspecialized
```
Comment (gaogaotiantian) [inline Lib/test/test_monitoring.py]: Maybe it's not a big deal, but I'm a little bit lost here about how this makes the specialization fail, and which specialization. Also I was a bit afraid when I saw "fail" in a test case. I thought you are going to fail the test on purpose. Maybe we can be a little bit more specific on what we are trying to do here?

**Review by gaogaotiantian — APPROVED** [review]

**Comment by brandtbucher** [comment]
@markshannon, this changed test is failing on tier two builds:

```pytb
Traceback (most recent call last):
  File "/Users/brandtbucher/cpython/Lib/test/test_monitoring.py", line 864, in test_implicit_stop_iteration
    self.check_events(implicit_stop_iteration, expected, recorders=recorders)
    ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brandtbucher/cpython/Lib/test/test_monitoring.py", line 749, in check_events
    self.assertEqual(events, expected)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
AssertionError: Lists differ: [] != [('raise', <class 'StopIteration'>)]
```

When events are set globally, we correctly invalidate all executors. However, we don't prevent the creation and entry of new ones. This patch fixes the issue:

```diff
diff --git a/Python/optimizer.c b/Python/optimizer.c
index f0793b8c8f2..433e04b382e 100644
--- a/Python/optimizer.c
+++ b/Python/optimizer.c
@@ -181,6 +181,9 @@ _PyOptimizer_Optimize(
     PyCodeObject *code = _PyFrame_GetCode(frame);
     assert(PyCode_Check(code));
     PyInterpreterState *interp = _PyInterpreterState_GET();
+    if (_PyCode_HAS_INSTRUMENTATION(code)) {
+        return 0;
+    }
     if (!has_space_for_executor(code, start)) {
         return 0;
     }
```

I'm not sure, but I *think* that we should also be invalidating some executors when local events are set. It's not needed to fix the test, just something that I noticed while poking around the instrumentation code. Something like this?

```diff
diff --git a/Python/instrumentation.c b/Python/instrumentation.c
index ae790a1441b..a4269055777 100644
--- a/Python/instrumentation.c
+++ b/Python/instrumentation.c
@@ -1990,6 +1990,9 @@ _PyMonitoring_SetLocalEvents(PyCodeObject *code, int tool_id, _PyMonitoringEvent
         goto done;
     }
     set_local_events(local, tool_id, events);
+#ifdef _Py_TIER2
+    _Py_Executors_InvalidateDependency(interp, code, 1);
+#endif
 
     res = force_instrument_lock_held(code, interp);
 ```

**Comment by brandtbucher** [comment]
GH-122413

**Pull request merged**